### PR TITLE
Add a "Preferences" window and rebindable hotkeys

### PR DESCRIPTION
**Pull request by AgmasGold**
This PR will add a preferences window, along with a way to rebind hotkeys (and actually see the keybindings available in the designer). 

~There are WPF `KeyBinding`, `KeyGesture` and `MouseGesture` classes available, but the `KeyBinding` and `KeyGesture` classes do not allow binding single key hotkeys - they throw an exception - hence the custom implementation in `PolyGesture`. Additionally, I am not sure how we would make sure to call `InvalidateVisual()` if using the commands directly in XAML, or map them to a key to make them rebindable.~

~There is an `InputBinding` class used to bind an `InputGesture` (the base class for `MouseGesture` and `KeyGesture` to an `ICommand`, but when attempting to add a `PolyGesture`, WPF threw an error - expecting either a `KeyGesture` or a `MouseGesture`, hence the need to implement my own class that derives from `InputBinding`.~ No longer needed. See comments below for reasoning.

~EDIT: This PR is now dependent on PR #206 (Localization data binding), due to the number of new localizations added within the new windows.~ (#206 has been merged).

* [x] Implement HotkeyCommandManager
* [x] Write tests for HotkeyCommandManager
* [x] Implement ManageKeybindingsViewModel
* [x] Create UI for ManageKeybindings.xaml
* [x] Refactor hardcoded commands in AnnoCanvas
  * [x] Key commands (copy, paste, rotate, delete)
* [x] Load Hotkeys from Settings
  * [x] Serialize to Settings
  * [x] Deserialize from Settings
* [x] Localize all the categories and settings
  * [x] Manage Keybindings localized
* [x] Update `SerializationHelper` - required to be able to serialize `Type`s
  * [x] Move to Newtonsoft.Json
  * [x] Update Implementation
* [x] Modify MainWindow UI
  * [x] Create Tools Menu Item
  * [x] Create Tools -> Preferences... Menu Item
  * [x] Remove "Extras" Menu Item and move under "Tools"
  * [x] Localize "Tools", "Preferences"
* [x] License window for ManageKeybindings page icons
  * [x] Window Created
  * [x] ViewModel Created
  * [x] Localize Licenses Window
* [X] Write tests
  * [x] Licenses window tests (if required, may be quite simple)
  * [x] Settings tests (loading/saving hotkeys)
* [ ] When closing the preferences window, the main window is put to the back of all other windows
* [ ] icons shown in Manage Keybindings are not always refreshed after closing the record window

This PR is already huge, with that in mind I'm going to leave the following ideas for separate PRs:
* Create UI and models for other settings categories (Updates, Extras etc.).
* Use mouse images in the window for recording hotkeys.
* Refactoring mouse commands out of Anno Canvas - this is probably quite a large job to undertake. Specifically, it could be useful to separate out zoom (mouse wheel), and copy item (middle click). The default MouseBinding class only uses a `MouseAction`, which does not have handling for mouse wheel scrolling.
* Multiple keybindings - allow the user to choose from a list of commands and add multiple hotkey configurations per command.

### CODE REVIEW

**Comment by FroggieFrog** [comment]
I have not checked it in detail, but I would propose to combine it with the introduction of a tabbed interface.
At first there would be one tab with the canvas and one tab with the settings.
Later it would be possible to have more tabs with more canvas open. So a user could edit multiple layouts. This was a request/discussion in the past.

**Comment by AgmasGold** [comment]
That would be an even more major change - probably would want to go into a separate PR. The only things I'm moving about from the main window is a few menu items.

**Comment by FroggieFrog** [comment]
Ok, maybe you are right. I will add a new issue for the tabbed interface.

You say there are erros when binding to a single key. I made a test project and can bind to everything I want.
**Can you explain why the code below is not possible for the designer?**

```xaml
    <Window.InputBindings>
        <!--Binding to single key-->
        <KeyBinding Key="F5"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed F5" />
        <KeyBinding Key="R"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed the key R" />
        <!--Binding to key + modifier-->
        <KeyBinding Key="F5"
                    Modifiers="Ctrl"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed Ctrl+F5" />
        <!--Binding to multiple modifiers-->
        <KeyBinding Key="F5"
                    Modifiers="Ctrl+Shift"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed Ctrl+Shift+F5" />
        <!--you can also use "Gesture" to combine Key + Modifiers in a string-->
        <KeyBinding Gesture="Control+Shift+F2"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed Ctrl+Shift+F2" />
        <!--The same is true for bindings to the mouse-->
        <MouseBinding  MouseAction="LeftDoubleClick"
                       Command="{Binding ShowMessageCommand}"
                       CommandParameter="You double clicked the left mouse button" />
        <MouseBinding  Gesture="Control+RightClick"
                       Command="{Binding ShowMessageCommand}"
                       CommandParameter="You clicked the right mouse button and pressed Ctrl" />
    </Window.InputBindings>
```

```csharp
    public class MessageCommand : ICommand
    {
        public void Execute(object parameter)
        {
            MessageBox.Show(parameter.ToString());
        }

        public bool CanExecute(object parameter)
        {
            return true;
        }

        public event EventHandler CanExecuteChanged;
    }
```

**Comment by AgmasGold** [comment]
2 things - 

- Try binding to something like R, or Shift + any letter.
- How we call `InvalidateVisual()` if binding like that - without calling `InvalidateVisual`, the canvas feels sluggish if we're calling something like `Rotate` (although it will update once the mouse is moved).

**Comment by FroggieFrog** [comment]
- I did a binding to the key `R` (second example)
- The call to `InvalidateVisual()` would be done by the class/ViewModel which contains the command. Maybe this class then needs a reference to `IAnnoCanvas` or a new interface which only contains the method `InvalidateVisual()`.
So no matter who calls the command, the logic would be the same and separated from the UI.
There could be separate classes for each command and the command does not care who executes it, be it a key or a mouse button or a key + modifier.
I think the main difficult is where to add/bind those commands. To the main window, the canvas or somewhere else?
- The remap could be done by offering several `ComboBox`es with the available `Key`s and `Modifier`s. So the user selects them to his preference and then the binding would be done via code. So maybe the Views need an interface with a method like `AddCommandBinding`.
Then there would be the separation of UI and logic again.
Something that has a reference to these interfaces and the concrete implementations with their ViewModels would also be needed.

I think this needs some more thinking and discussion.
But in my opinion there is no need for the custom `PolyGesture` and `PolyBinding`. At least at the moment I don't see their value.

**Comment by AgmasGold** [comment]
Yep - you're right - no issues binding to a single key - although weirdly the R key binding doesn't work when the XAML code is moved into the MainWindow, but does work from AnnoCanvas - so not a major issue, just something in MainWindow that must be capturing and handling the single key keypresses.

One more thing I've thought about - how do we know from the code which binding represents which command? 

My current idea is that we'd need to create the input bindings through code, rather than in the XAML, using the `AddCommandBinding` system or whatever we end up going with using information from settings. That then means we can then map the instance of the binding to a key, and go from there.

Apparently we can't create single key keybindings from code, so that must have been where I got my information from:
`new KeyGesture(Key.R, ModifierKeys.None);` throws
`NotSupportedException: 'None+R' key and modifier combination is not supported for KeyGesture.`

**Comment by AgmasGold** [comment]
From https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Input/Command/KeyGesture.cs,208

All normal and digit keys with no modifiers are apparently invalid:

```csharp
///<summary>
/// Is Valid Keyboard input to process for commands
///</summary>
internal static bool IsValid(Key key, ModifierKeys modifiers)
{
    //
    //  Don't enforce any rules on the Function keys or on the number pad keys.
    //
    if (!((key >= Key.F1 && key <= Key.F24) || (key >= Key.NumPad0 && key <= Key.Divide)))
    {
        //
        //  We check whether Control/Alt/Windows key is down for modifiers. We don't check
        //  for shift at this time as Shift with any combination is already covered in above check.
        //  Shift alone as modifier case, we defer to the next condition to avoid conflicing with
        //  TextInput.

        if ((modifiers & (ModifierKeys.Control | ModifierKeys.Alt | ModifierKeys.Windows)) != 0)
        {
            switch (key)
            {
                case Key.LeftCtrl:
                case Key.RightCtrl:
                case Key.LeftAlt:
                case Key.RightAlt:
                case Key.LWin:
                case Key.RWin:
                    return false;

                default:
                    return true;
            }
        }
        else if ((key >= Key.D0 && key <= Key.D9) || (key >= Key.A && key <= Key.Z))
        {
            return false;
        }
    }
    return true;
}
```

**Comment by AgmasGold** [comment]
XAML:
```xaml
<UserControl x:Class="AnnoDesigner.AnnoCanvas"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008" xmlns:models="clr-namespace:AnnoDesigner.Models"
             mc:Ignorable="d"
             d:DesignHeight="200"
             d:DesignWidth="200"
             Focusable="True"
             ClipToBounds="True">
    <UserControl.InputBindings>
        <!--Binding to single key-->
        <KeyBinding  Key="R"
                    Command="{Binding ShowMessageCommand}"
                    CommandParameter="You pressed R">
        </KeyBinding>
    </UserControl.InputBindings>
    <Grid Name="StatisticsPanel"
          HorizontalAlignment="Right"
          Background="White">

    </Grid>
</UserControl>
```
Actual created gesture:
image

Code from https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Input/Command/KeyGestureConverter.cs,88

image

**Comment by FroggieFrog** [comment]
So `KeyGesture` is a little strange. Therefore don't use it from code behind.
Following does work:

```csharp
    public partial class MainWindow : Window
    {
        private MainViewModel _viewModel;

        public MainWindow()
        {
            InitializeComponent();
        }

        private void Window_Loaded(object sender, RoutedEventArgs e)
        {
            _viewModel = DataContext as MainViewModel;

            //exception -> do not use Gesture
            //InputBindings.Add(new KeyBinding(_viewModel.ShowMessageCommand, new KeyGesture(Key.D, ModifierKeys.None)));

            //exception -> do not use Gesture
            //InputBindings.Add(new KeyBinding(_viewModel.ShowMessageCommand, Key.D, ModifierKeys.None));

            //exception -> do not use Gesture
            //InputBindings.Add(new InputBinding(_viewModel.ShowMessageCommand, new KeyGesture(Key.D, ModifierKeys.None)));

            //binding to sinlge key
            InputBindings.Add(new KeyBinding
            {
                Command = _viewModel.ShowMessageCommand,
                CommandParameter = "you pressed the key 'D'",
                Key = Key.D
            });

            //binding to sinlge key with multiple modifiers
            InputBindings.Add(new KeyBinding
            {
                Command = _viewModel.ShowMessageCommand,
                CommandParameter = "you pressed Ctrl + Alt + the key 'D'",
                Key = Key.D,
                Modifiers = ModifierKeys.Control | ModifierKeys.Alt
            });

            //Modifiers.None works when set as property
            InputBindings.Add(new KeyBinding
            {
                Command = _viewModel.ShowMessageCommand,
                CommandParameter = "you pressed the key 'F'",
                Key = Key.F,
                Modifiers = ModifierKeys.None
            });

            //MouseBinding works also
            InputBindings.Add(new MouseBinding
            {
                Command = _viewModel.ShowMessageCommand,
                CommandParameter = "you clicked the middle mouse button",
                MouseAction = MouseAction.MiddleClick
            });
        }
    }
```

**Comment by AgmasGold** [comment]
Cool - we'll go with  that then, and I'll remove the `PolyBinding` and `PolyGesture` classes. I might just add a comment with a link to this comment thread as an explanation as to why we have to create them that way.

File: AnnoDesigner/KeyRecorderWindow.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+using System.Windows.Media.Imaging;
+using System.Windows.Shapes;
+using Microsoft.Win32.SafeHandles;
+
+namespace AnnoDesigner
+{
+    /// <summary>
+    /// Interaction logic for KeyRecorderWindow.xaml
+    /// </summary>
+    public partial class HotkeyRecordingWindow : Window
```
Comment (FroggieFrog) [inline AnnoDesigner/KeyRecorderWindow.xaml.cs:24]: classname does not match filename

File: AnnoDesigner/KeyRecorderWindow.xaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿<Window x:Class="AnnoDesigner.HotkeyRecordingWindow"
```
Comment (FroggieFrog) [inline AnnoDesigner/KeyRecorderWindow.xaml:1]: - classname does not match filename
- File has a lot of code behind. Maybe extract to a ViewModel?

Comment (AgmasGold) [reply AnnoDesigner/KeyRecorderWindow.xaml]: Yeah, this was a proof of concept. I will move it to either a ViewModel or possibly its own UserControl in future.

File: AnnoDesigner/KeyRecorderWindow.xaml.cs (hunk 1 above)
Comment (AgmasGold) [reply AnnoDesigner/KeyRecorderWindow.xaml.cs]: Fixed.

File: AnnoDesigner/KeyRecorderWindow.xaml (hunk 2 above)
Comment (FroggieFrog) [reply AnnoDesigner/KeyRecorderWindow.xaml]: Ok, it was just "to note it".

**Comment by AgmasGold** [comment]
The `LicensesWindow` has been completed. I've included some screenshots below. Everything should be fully localized where needed.
<details>
<summary>View images</summary>

image
image
image
The chevron animates between states (it took me way too long to add this in Blend, but it looks nice)
image

</details>

**Comment by AgmasGold** [comment]
I'll get those changes added in. 

As the to Preferences window missing an OK/Save button, as we don't have a way of rolling the changed settings back, I guess I could add a `Close` button? I just think that could cause confusion, especially if certain settings pages might have a rollback feature in future (so, for example, could have an `Apply`/`Save` button and a `Cancel` button.

The `Update Settings` page I will just delete - moving other settings to the window will be a separate PR, as this one already so huge.

**Comment by FroggieFrog** [comment]
A `Close` button would be sufficient.
Maybe don't add it to the preferences window, but to the page as a workaround?

**Comment by AgmasGold** [comment]
I don't think thats possible, unless I add the close command as another depedency to the view model. I guess maybe the best solution given the situation would be to put the close button button beneath the categories list - a little bit of an odd place, but then it will not interfere with anything added to a page, and responsibility of closing the window stays with window.

**Comment by FroggieFrog** [comment]
- [ ]  Is it possible to select the first (and currently only one) ListViewItem/Page by default?
So opening the Preferences window directly displays the `ManageKeyBindingsPage`.
- [ ] When closing the preferences window, the main window is put to the back of all other windows.
  - open Explorer
  - open AnnoDesigner (above Explorer window)
  - open Preferences
  - switch to Explorer
  - close Preferences -> Explorer is shown
- [ ] Is it possible to fix the start/default position of the preferences window? Like always open in the center of the app?
- [ ] Should the preferences window made modal?

**Comment by AgmasGold** [comment]
* I'll look into selecting the first item by default. It might just be a question of manually doing it in the view model, or there might be a way to do it by reading the first `ListViewItem`.
* Not sure why this is occuring, I'll investigate. I'm not doing anything different to other windows, I'm just calling `Close()`.
* It should be opening in the center of the app already.
* Not sure about this - I think my initial thought was that if we eventually put other settings in there there may be a case where you might want to check something on the MainWindow (shortcut, tweak in functionality etc.), which would not be possible if this was made Modal.

**Comment by AgmasGold** [comment]
Just looking at the middle-click issue. Its not the command does not refresh, its that MiddleClick is hardcoded to clone and rotate an object within AnnoCanvas.xaml.cs. I could remove that functionality, but then that would be a breaking change, and could disrupt users' experiences within the app. 

For now, I propose to leave it, as refactoring the hardcoded mouse commands in AnnoCanvas is a major piece of work, and is currently outside the scope of this PR.

**Pull request merged**